### PR TITLE
Feat: Localisation zone system creation

### PR DESCRIPTION
**Pull request by amber-mertens-tfn**
# Changes

Added a new standalone script (localisation.py) to create a new combined zoning system for localisation. 

Used and adapted function produce_zoning in comp_zone.py, but left the original there (not sure how to consolidate this/where this functionality should end up).

# Tasks

- [ ] Have unittests been added (testing individual functions and their edge cases)
- [ ] Has documentation (including user guide) been updated
- [ ] Have new dependencies been added (or changed) in relevant `requirements.txt`
- [X] Code linting, tests and other workflows pass

### CODE REVIEW

File: src/scripts/localisation.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    zones = zones.rename(columns={id_col: "id"})
+    zones["zoning"] = zone_name
+
+    return zones
+
+
+def produce_zoning(
+    ext_zones: ZoneSystemInfo,
+    int_zones: ZoneSystemInfo,
+    buff_zones: ZoneSystemInfo | None,
```
Comment (isaac-tfn) [inline src/scripts/localisation.py]: Where None is an option it is usually better to have it default to None, so that the function can be called easily without passing those arguments at all

File: src/scripts/localisation.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            parameters.zone_systems.buffer_zones,
+            bound,
+            buffer_bound,
+        )
+
+        new_zones.to_file(
+            parameters.output_folder
+            / (
+                f"zoning_localisation_{parameters.localisation_area.area_name}_"
+                f"{parameters.zone_systems.internal_zones.name}.shp"
```
Comment (isaac-tfn) [inline src/scripts/localisation.py]: May want an option for output file format (e.g. .shp, .gpkg)

**Review by isaac-tfn — COMMENTED** [review]
Sorry this has taken so long to review!

It looks good, the only thing I would say is there probably needs to be more checks for compatibility of zone systems. At the moment I think depending on what was chosen, int/buffer/ext zones could overlap each other. The easiest thing to do would be to actually trim the inner zones to the inner bound, buffer zones to buffer bound etc., and raise a warning for the user that this is happening.

It could also be useful to be able to include the 'master zone system', which in our case is lsoa 2021 as that is the zone system out land use data uses, and get some info about the compatibility of the created zone system with that, like again whether it nests properly.

**Comment by amber-mertens-tfn** [comment]
Changed the way the new zones get selected:
Spatial translation between zoning system and boundary: assign zones to that boundary zone with which they have the largest overlap (if the boundary consists of several zones) and keep only zones that exceed the overlap threshold (0.5 default). 
Separated steps into functions (used code clean up suggestions by copilot).

File: src/scripts/localisation.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    normalized = output_format.casefold()
+
+    if normalized in _SHAPEFILE_FORMATS:
+        return "ESRI Shapefile", "shp"
+
+    if normalized in _GPKG_FORMATS:
+        return "GPKG", "gpkg"
+
+    LOG.warning("Output format %s not recognised, defaulting to geopackage.", output_format)
+    return "GPKG", "gpkg"
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py]: I think you could reimplement this functionality within a FileFormat enum and then use that as a type annotation for the config so that the format is validated at input.

The enum could be defined as below with:
- `_missing_` method to have the flexibility of passing different strings to the config
- `driver` and `suffix` properties could replace this function (`get_output_driver_and_extension`)

```python
class FileFormat(enum.StrEnum):
    """GIS file formats."""

    GEOPACKAGE = enum.auto()
    SHAPEFILE = enum.auto()

    @classmethod
    def _missing_(cls, value) -> "FileFormat":
        """Case insensitive and more flexible strings accepted."""
        value = str(value).strip().lower()
        for i in cls:
            if value == i.value:
                return i
        
        if value in _SHAPEFILE_FORMATS:
            return cls.SHAPEFILE
        if value in _GPKG_FORMATS:
            return cls.GEOPACKAGE
        return None

    @property
    def driver(self) -> str:
        """File format IO driver."""
        lookup = {FileFormat.GEOPACKAGE: "GPKG", FileFormat.SHAPEFILE: "ESRI Shapefile"}
        return lookup[self]

    @property
    def suffix(self) -> str:
        """Path suffix (extension) for this format."""
        lookup = {FileFormat.GEOPACKAGE: "gpkg", FileFormat.SHAPEFILE: "shp"}
        return lookup[self]
```

The class above could even be defined within caf.space so it can be used more widely.

File: src/scripts/localisation.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+    external_zones: ZoneSystemInfo
+    buffer_zones: ZoneSystemInfo
+    target_zones: ZoneSystemInfo
+
+
+class _Config(ctk.BaseConfig):
+    """Config for running localisation zoning script."""
+
+    output_path: pydantic.DirectoryPath
+    output_format: str
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py]: If you implemented `FileFormat` enum ([comment below](https://github.com/Transport-for-the-North/caf.space/pull/63#discussion_r3225772913)) this could be used to validate the config
```suggestion
    output_format: FileFormat
```

File: src/scripts/localisation.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+        boundary_zones.shapefile,
+        columns=[
+            boundary_zones.id_col,
+            selected_area.selected_colname,
+        ],
+    )
+    bound = bound_zones[
+        bound_zones[selected_area.selected_colname].isin(selected_area.selected_zones)
+    ]
+    buffer_bound = bound_zones[bound_zones.geometry.touches(bound.union_all())]
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py]: Is touches a little bit restrictive and likely to miss things? The touches definition is:
> An object is said to touch other if it has at least one point in common with other and its interior does not intersect with any part of the other. Overlapping features therefore do not touch.

If this hasn't caused an issue in the tests done already that it's probably fine to leave as is for now but if there are issues with other zone systems in future this could be updated to be a bit more robust e.g. checks intersects but not within and perhaps includes a small buffer around bound.

File: src/scripts/localisation.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            "%s zone(s) were dropped because their overlap with the boundary was lower than the overlap threshold (%s). "
+            "Their IDs are: %s and the maximum overlap is %s",
+            len(dropped_zones),
+            overlap_threshold,
+            (", ".join(str(zoneid) for zoneid in dropped_zones[id_col].values)),
+            round(dropped_zones[factor_col].max(), 3),
+        )
+
+    # Join the translation factors to the zone geometries
+    zone_gdf = gpd.read_file(zone_system.shapefile, columns=[zone_system.id_col, "geometry"])
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py:235]: Not to fix now but would be better to not read the shapefile twice (once within `ZoneTranslation` above), issue #57 talks about restructing the translation to make accessing the loaded data easier and cache read values.

File: src/scripts/localisation.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+    return TransZoneSystemInfo(**data)
+
+
+def write_boundary_files(
+    output_folder: pathlib.Path,
+    internal_bound: gpd.GeoDataFrame,
+    buffer_bound: gpd.GeoDataFrame,
+    driver: str,
+    extension: str,
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py]: Pass a `FileFormat` value instead of these two, see [comment](https://github.com/Transport-for-the-North/caf.space/pull/63#discussion_r3225772913) above.

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Logic all looks good just one question about using touches.

There's a few comments all suggesting adding an enum to define the file format and passing that around instead of just using a string.

File: src/scripts/localisation.py [hunk 8]
```diff
@@ -0,0 +1,676 @@
...
+
+    if bound.empty:
+        raise ValueError(
+            "No boundary zones were selected for localisation. "
+            f"Check selected zones {selected_area.selected_zones} for "
+            f"column name {selected_area.selected_colname} in {boundary_zones.name} "
+            f" with shapefile at {boundary_zones.shapefile}."
+        )
+
+    # buffer_bound_touches = bound_zones[bound_zones.geometry.touches(bound.union_all())]
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py]: This commented out code can be removed if we've switched to intersects below
```suggestion
```

File: src/scripts/localisation.py [hunk 9]
```diff
@@ -0,0 +1,676 @@
...
+                f"{from_name}_to_{to_name}": f"{from_name}_to_{to_name}_{translation_type}",
+                f"{to_name}_to_{from_name}": f"{to_name}_to_{from_name}_{translation_type}",
+            }
+        )
+        rounded_non_matched_parts.append(rounded)
+
+    lookup_rounded = functools.reduce(
+        lambda left, right: left.merge(right, on=id_cols, how="outer"),
+        rounded_non_matched_parts,
+    )
```
Comment (MattBuckley-TfN) [inline src/scripts/localisation.py]: Instead of calling merge multiple times you should be able to set the index of `rounded` to the `id_cols` and then use `pd.concat` to combine all the DataFrames at once instead of iterating through them.

**Review by MattBuckley-TfN — CHANGES_REQUESTED** [review]
Changes look good just a couple of comments.

**Review by MattBuckley-TfN — APPROVED** [review]
Looks great
